### PR TITLE
Reduce the use of C++ exceptions

### PR DESCRIPTION
**Pull request by mdboom**
## PR Summary

This is hopefully the last of the pyodide-related patches for a while.

Emscripten has only limited support for C++ exceptions.  The support that is there doesn't seem to work fully correctly over the matplotlib code base, and even if it did work, it has a large performance penalty which will be hard to resolve.  The best available option at the moment is to have C++ exception support turned off, which means a `throw` from C++ aborts the current WebAssembly frame, exiting back to Javascript without a chance for any C++ `catch` blocks to handle the exception.  (This is weaker than an unhandled exception in "native" code which would terminate the application altogether).  In most cases, this behavior is "good enough" when the exceptions are truly "exceptional".  It's not ideal, since there isn't an opportunity to cleanup or provide friendly error messages from C++.

However there is a very small handful of places in matplotlib where exception-handling is used for cases that are the normal course of business, such as dealing with an input differently based on the number of its dimensions.  This grew out of the fact that a constructor can not return an error code, so must raise an exception to indicate a dimension mismatch.  I think it's worth fixing these three places to just use the C-level Numpy API instead of the `numpy_cpp.h` abstraction, even though that's not strictly necessary for most platforms.  There should be performance benefits for everyone, however, since it avoids interpreting the array-like input twice.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: src/_path_wrapper.cpp [hunk 1]
```diff
@@ -437,23 +437,25 @@ static PyObject *Py_affine_transform(PyObject *self, PyObject *args, PyObject *k
...
+    }
+
+    if (PyArray_NDIM(vertices_arr) == 2) {
+        numpy::array_view<double, 2> vertices(vertices_arr);
         npy_intp dims[] = { (npy_intp)vertices.size(), 2 };
         numpy::array_view<double, 2> result(dims);
         CALL_CPP("affine_transform", (affine_transform_2d(vertices, trans, result)));
+        Py_DECREF(vertices_arr);
+        return result.pyobj();
+    } else { // PyArray_NDIM(vertices_arr) == 1
```
Comment (jkseppan) [inline src/_path_wrapper.cpp]: Do both of these branches get executed in tests?

File: src/py_converters.cpp [hunk 2]
```diff
@@ -158,45 +158,36 @@ int convert_rect(PyObject *rectobj, void *rectp)
...
-                }
-
-                rect->x1 = rect_arr(0);
-                rect->y1 = rect_arr(1);
-                rect->x2 = rect_arr(2);
-                rect->y2 = rect_arr(3);
-            }
-            catch (py::exception &)
-            {
+        } else {  // PyArray_NDIM(rect_arr) == 1
```
Comment (jkseppan) [inline src/py_converters.cpp]: Same question about test coverage.

File: src/_path_wrapper.cpp [hunk 3]
```diff
@@ -437,23 +437,25 @@ static PyObject *Py_affine_transform(PyObject *self, PyObject *args, PyObject *k
...
+    if (vertices_arr == NULL) {
+        return NULL;
+    }
+
+    if (PyArray_NDIM(vertices_arr) == 2) {
+        numpy::array_view<double, 2> vertices(vertices_arr);
         npy_intp dims[] = { (npy_intp)vertices.size(), 2 };
         numpy::array_view<double, 2> result(dims);
         CALL_CPP("affine_transform", (affine_transform_2d(vertices, trans, result)));
+        Py_DECREF(vertices_arr);
```
Comment (anntzer) [inline src/_path_wrapper.cpp]: Why was Py_DECREF unneeded before?

**Comment by mdboom** [comment]
That's a good question about coverage.  I just confirm that all of those branches are exercised by the existing tests (just crudely by adding `printf` statements).  Longer term, we should probably do coverage testing on the C++ which I don't think we currently do.

**Pull request merged**